squad_client/tux.py

```diff
@@ -0,0 +1,86 @@
+import hashlib
+import json
+import logging
+import os
+import posixpath
+import tempfile
+import urllib
+
+
+logger = logging.getLogger(__name__)
+
+
+ALLOWED_METADATA = [
+    "config",
+    "download_url",
+    "git_branch",
+    "git_commit",
+    "git_describe",
+    "git_ref",
+    "git_repo",
+    "git_sha",
+    "git_short_log",
+    "kconfig",
+    "kernel_version",
+    "make_kernelversion",
+]
+
+
+def load_builds(build_json):
+    builds = None
+
+    try:
+        with open(build_json) as fp:
+            builds = json.load(fp)
+
+    except json.JSONDecodeError as jde:
+        logger.error("Failed to load json: %s", jde)
+
+    except OSError as ose:
+        logger.error("Failed to open file: %s", ose)
+
+    return builds
+
+
+def build_test_name(build):
+    sha = combined_kconfig_sha(build)
+
+    return "build/%s-%s-%s" % (
+        build["build_name"],
+        build["kconfig"][0],
+        sha
+    )
+
+
+def buildset_test_name(build):
+    return "build/%s-%s-%s-%s" % (
+        build["build_name"],
+        build["target_arch"],
+        build["toolchain"],
+        build["kconfig"][0],
+    )
+
+
+def combined_kconfig_sha(build):
+    tmp_dir = tempfile.mkdtemp()
+    kconfig = os.path.join(tmp_dir, "kconfig")
+    with open(kconfig, "w") as kfp:
+        for k in build["kconfig"][1:]:
+            url = urllib.parse.urlparse(k)
+
+            if url.scheme in ["http", "https"]:
+                partial = os.path.basename(url.path)
+                src = urllib.parse.urljoin(build["download_url"], posixpath.join("fragments", partial))
+                dst = os.path.join(tmp_dir, partial)
+                urllib.request.urlretrieve(src, dst)
```

Comment:
I'll be a bit pickier in this one :)

1. You can use `requests.get()` instead of `urllib.request.urlretrieve()`. This way you don't need any temp files, you could just use `requests.get().text` to get the contents. Here's some discussion about it: https://gist.github.com/kennethreitz/973705

2. It seems you're now fetching the actual kconfig file from the URL, instead of just using the URL to make up the sha.
I think it would be nice to actually test this download, you know, just to make sure. You could do that by mocking `requests.get` by a function that returns certain contents depending on the arg you give to `requests.get`
I did this in squad to test from for [Gerrit plugin test](https://github.com/Linaro/squad/blob/master/test/plugins/test_gerrit.py#L153).

Reply:
@chaws I'll ping you Friday and we'll chat about this one.